src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/OkHttpWebSocketWrapper.java

```diff
@@ -97,39 +102,66 @@ public void onMessage(WebSocket webSocket, ByteString bytes) {
 
         @Override
         public void onClosing(WebSocket webSocket, int code, String reason) {
-            onClose.invoke(code, reason);
+            boolean isOpen = false;
+            stateLock.lock();
+            try {
+                isOpen = startSubject.hasComplete();
+            } finally {
+                stateLock.unlock();
+            }
+            // Only call onClose if connection is open
+            if (isOpen) {
+                onClose.invoke(code, reason);
+            }
+
             try {
-                closeLock.lock();
+                stateLock.lock();
                 closeSubject.onComplete();
             }
             finally {
-                closeLock.unlock();
+                stateLock.unlock();
             }
-            checkStartFailure();
+            checkStartFailure(null);
         }
 
         @Override
         public void onFailure(WebSocket webSocket, Throwable t, Response response) {
             logger.error("WebSocket closed from an error: {}.", t.getMessage());
 
+            boolean isOpen = false;
             try {
-                closeLock.lock();
+                stateLock.lock();
                 if (!closeSubject.hasComplete()) {
                     closeSubject.onError(new RuntimeException(t));
                 }
+
+                isOpen = startSubject.hasComplete();
             }
             finally {
-                closeLock.unlock();
+                stateLock.unlock();
+            }
+            // Only call onClose if connection is open
+            if (isOpen) {
+                onClose.invoke(null, t.getMessage());
             }
-            onClose.invoke(null, t.getMessage());
-            checkStartFailure();
+            checkStartFailure(t);
         }
 
-        private void checkStartFailure() {
-            // If the start task hasn't completed yet, then we need to complete it
-            // exceptionally.
-            if (!startSubject.hasComplete()) {
-                startSubject.onError(new RuntimeException("There was an error starting the WebSocket transport."));
+        private void checkStartFailure(Throwable t) {
+            stateLock.lock();
+            try {
+                // If the start task hasn't completed yet, then we need to complete it
+                // exceptionally.
+                if (!startSubject.hasComplete()) {
+                    String errorMessage = "There was an error starting the WebSocket transport. ";
+                    if (t != null) {
+                        startSubject.onError(new RuntimeException(errorMessage + t.getMessage()));
+                    } else {
+                        startSubject.onError(new RuntimeException(errorMessage));
+                    }
```

Comment:
Why not the following?

```suggestion
                    startSubject.onError(new RuntimeException("There was an error starting the WebSocket transport.", t));
```

Reply:
... I could have sworn I tried that and didn't see any overloads. Maybe my intellisense hadn't updated yet.